# Proyect 7 using Streamlit 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import math
from scipy import stats as st

In [2]:
# Carga de archivos de DataFrame

print("#############################################")
print("#                VEHICLES-US                #")
print("#############################################")

print("\n╔════════════════════════════════╗")  
print("║   ANÁLISIS DE CIENCIA DE DATOS ║")  
print("║         DESARROLLADO POR       ║")  
print("║    Mauricio Zárate Romero      ║")  
print("╚════════════════════════════════╝") 

# car_data = df_cd

df_cd_copy = pd.read_csv('vehicles_us_limpio.csv')

# Copia de seguridad para no modificar los DataFrames originales.

df_cd = df_cd_copy.copy()

#############################################
#                VEHICLES-US                #
#############################################

╔════════════════════════════════╗
║   ANÁLISIS DE CIENCIA DE DATOS ║
║         DESARROLLADO POR       ║
║    Mauricio Zárate Romero      ║
╚════════════════════════════════╝


In [3]:
# Información general/resumida sobre el DataFrame

print("\n╔══════════════════════════════════════════════════╗")
print("║                'vehicles_us.csv'                 ║")
print("║                  'Información'                   ║")
print("╚══════════════════════════════════════════════════╝")
print("\n" + "=" * 50)
df_cd.info()
print("\n" + "=" * 50)
print("\n- Total de filas duplicadas:", df_cd.duplicated().sum())
print("\n" + "=" * 50)
print('\n- Total de valores nulos:\n \n', df_cd.isna().sum())
print("\n" + "=" * 50)
print('\n- Estadisticas\n \n', df_cd.describe())


╔══════════════════════════════════════════════════╗
║                'vehicles_us.csv'                 ║
║                  'Información'                   ║
╚══════════════════════════════════════════════════╝

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35622 entries, 0 to 35621
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         35622 non-null  int64  
 1   model_year    35622 non-null  int64  
 2   model         35622 non-null  object 
 3   cylinders     35622 non-null  int64  
 4   fuel          35622 non-null  object 
 5   odometer      35622 non-null  float64
 6   transmission  35622 non-null  object 
 7   type          35622 non-null  object 
 8   paint_color   35622 non-null  object 
 9   is_4wd        35622 non-null  int64  
 10  date_posted   35622 non-null  object 
 11  days_listed   35622 non-null  int64  
 12  condition     35622 non-null  object 
dtypes: float64(1), int64(5), ob

In [4]:
df_cd.sample(10)

price  model_year                       model  cylinders    fuel  \
6914   28900        2011               ford f-350 sd          8  diesel   
16856   9985        2012               buick enclave          6     gas   
29634  10977        2015                nissan rogue          4     gas   
32695   5000        2005                    ram 1500          8     gas   
32328   7995        2009             gmc sierra 1500          8     gas   
21430   4999        2003             chevrolet tahoe          8     gas   
31153   3450        2012               nissan altima          4     gas   
13274  14500        2004  chevrolet silverado 2500hd          8     gas   
9870    8500        2016                  ford focus          4     gas   
18097   9995        2014                 ford escape          4     gas   

       odometer transmission    type paint_color  is_4wd date_posted  \
6914   120000.0    automatic   truck       white       1  2018-06-30   
16856  135030.0    automatic     SUV        grey       1  2018-05-22   
29634   66315.0    automatic     SUV        grey       0  2018-10-21   
32695  122000.0    automatic  pickup        grey       1  2019-02-26   
32328  158119.0    automatic  pickup        blue       1  2018-06-12   
21430  140000.0    automatic     SUV     Unknown       0  2019-02-05   
31153  170000.0    automatic   sedan     Unknown       0  2018-11-12   
13274  106000.0    automatic   truck        grey       1  2018-05-04   
9870    36237.0       manual   sedan       black       0  2018-12-31   
18097  131571.0    automatic     SUV       white       1  2018-08-28   

       days_listed  condition  
6914            12       fair  
16856           32       fair  
29634           32       good  
32695            5       fair  
32328           18       fair  
21430           42       fair  
31153           32       fair  
13274           32       fair  
9870            38  excellent  
18097           25       fair

### Tratamiento por columna

### price

In [5]:
# 1. Mostrar valores nulos y tipo de columna antes de modificar
print(f"Valores nulos = {df_cd['price'].isna().sum()}")
print(f"Tipo de columna = {df_cd['price'].dtypes}")

# 2. Reemplazar los valores 1 por NaN
indices_1 = df_cd[df_cd['price'] == 1].index.tolist()
df_cd.loc[indices_1, 'price'] = np.nan

# 3. Mostrar valores nulos y tipo de columna después de modificar
print(f"Valores nulos = {df_cd['price'].isna().sum()}")
print(f"Tipo de columna = {df_cd['price'].dtypes}")

# 4. Comprobar que ya no hay valores 1 en 'price'
indices_1 = df_cd[df_cd['price'] == 1].index.tolist()
print(f"Índices con price = 1: {indices_1}")

# 5. Eliminar filas con NaN en 'price' (opcional, si quieres limpiar el DataFrame)
df_cd = df_cd.dropna(subset=['price'])

# 6. Mostrar valores nulos tras eliminar filas
print(f"Valores nulos después de dropna = {df_cd['price'].isna().sum()}")

# 7. Cambiar la columna a int
df_cd['price'] = df_cd['price'].astype(int)
print(df_cd['price'].dtype)


Valores nulos = 0
Tipo de columna = int64
Valores nulos = 0
Tipo de columna = int64
Índices con price = 1: []
Valores nulos después de dropna = 0
int64


### model_year

In [6]:
# Mostrar estadísticas descriptivas de la columna 'model_year' antes de eliminar nulos
print(df_cd['model_year'].describe())

# Imprimir la cantidad de valores nulos en la columna 'model_year' antes de eliminarlos
print(f'\nvalores nulos = {df_cd["model_year"].isna().sum()}')

# Eliminar filas donde 'model_year' es nulo
df_cd = df_cd.dropna(subset=['model_year'])

# Imprimir la cantidad de valores nulos en la columna 'model_year' después de eliminarlos
print(f'\nvalores nulos despues de limpiarlos = {df_cd["model_year"].isna().sum()}')

df_cd['model_year'] = df_cd['model_year'].astype(int)

print(f'\nCambio a tipo de dato = {df_cd['model_year'].dtype}')


valores_ordenados_my = df_cd['model_year'].drop_duplicates().sort_values()

valores_ordenados_my = np.unique(df_cd['model_year'])
print(valores_ordenados_my)









count    35622.000000
mean      2009.664337
std          6.219318
min       1908.000000
25%       2006.000000
50%       2011.000000
75%       2014.000000
max       2019.000000
Name: model_year, dtype: float64

valores nulos = 0

valores nulos despues de limpiarlos = 0

Cambio a tipo de dato = int64
[1908 1936 1948 1955 1958 1960 1961 1962 1963 1964 1965 1966 1967 1968
 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978 1979 1980 1981 1982
 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994 1995 1996
 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009 2010
 2011 2012 2013 2014 2015 2016 2017 2018 2019]


### model

In [7]:
print(df_cd['model'].describe())

print(f'\nvalores nulos = {df_cd["model"].isna().sum()}')

valores_ordenados_m = np.sort(np.unique(df_cd['model']))
print('\n')
print(valores_ordenados_m)

print(f'\nEl tipo de columna es correcto = {df_cd['model'].dtype}')

count          35622
unique            99
top       ford f-150
freq            1924
Name: model, dtype: object

valores nulos = 0


['acura tl' 'bmw x5' 'buick enclave' 'cadillac escalade'
 'chevrolet camaro' 'chevrolet camaro lt coupe 2d' 'chevrolet colorado'
 'chevrolet corvette' 'chevrolet cruze' 'chevrolet equinox'
 'chevrolet impala' 'chevrolet malibu' 'chevrolet silverado'
 'chevrolet silverado 1500' 'chevrolet silverado 1500 crew'
 'chevrolet silverado 2500hd' 'chevrolet silverado 3500hd'
 'chevrolet suburban' 'chevrolet tahoe' 'chevrolet trailblazer'
 'chevrolet traverse' 'chrysler 200' 'chrysler 300'
 'chrysler town & country' 'dodge charger' 'dodge dakota'
 'dodge grand caravan' 'ford econoline' 'ford edge' 'ford escape'
 'ford expedition' 'ford explorer' 'ford f-150' 'ford f-250'
 'ford f-250 sd' 'ford f-250 super duty' 'ford f-350 sd' 'ford f150'
 'ford f150 supercrew cab xlt' 'ford f250' 'ford f250 super duty'
 'ford f350' 'ford f350 super duty' 'ford focus' 'ford focus se

Al parecer la columna no necesita tratamiento

### condition

In [8]:
print(df_cd['condition'].describe())

print(f'\nvalores nulos = {df_cd["condition"].isna().sum()}')

valores_ordenados_c = np.sort(np.unique(df_cd['condition']))
print('\n')
print(valores_ordenados_c)

# Filtrar filas donde 'condition' es 'new' y 'odometer' no es 0
condicion_new_odometro_no_cero = df_cd[(df_cd['condition'] == 'new') & (df_cd['odometer'] != 0)]
print(f"\nCantidad de autos 'new' con odómetro distinto de 0: {len(condicion_new_odometro_no_cero)}")

condicion_new_odometro_si_cero = df_cd[(df_cd['condition'] == 'new') & (df_cd['odometer'] == 0)]
print(f"\nCantidad de autos 'new' con odómetro distinto de 0: {len(condicion_new_odometro_si_cero)}")




print(f'\nEl tipo de columna es correcto = {df_cd['condition'].dtype}')

count     35622
unique        5
top        fair
freq      15833
Name: condition, dtype: object

valores nulos = 0


['excellent' 'fair' 'good' 'like new' 'salvage']

Cantidad de autos 'new' con odómetro distinto de 0: 0

Cantidad de autos 'new' con odómetro distinto de 0: 0

El tipo de columna es correcto = object


por lo que parece la columna de condition es erronea por lo que se le dara tratamiento estableciendo nuevos parametros para definir la condicion del vehiculo y quitando la columna de new ya que todos los vehiculos son usados y no deberia de haber datos con odometro en 0

In [9]:
def nuevas_condiciones(odometer):
    if pd.isna(odometer) or odometer == 0:
        return np.nan
    elif odometer <= 10000:
        return 'like new'
    elif odometer <= 40000:
        return 'excellent'
    elif odometer <= 100000:
        return 'good'
    elif odometer <= 180000:
        return 'fair'
    else:
        return 'salvage'
    
df_cd['condition_fixed'] = df_cd['odometer'].apply(nuevas_condiciones)


df_cd = df_cd.dropna(subset='condition_fixed').copy()


df_cd.drop(columns=['condition'], inplace=True)
df_cd.rename(columns={'condition_fixed' : 'condition'}, inplace=True)

print(df_cd['condition'].value_counts())

print('\n')

df_cd.info()

condition
fair         15833
good         10228
salvage       5274
excellent     3277
like new      1010
Name: count, dtype: int64


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35622 entries, 0 to 35621
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         35622 non-null  int64  
 1   model_year    35622 non-null  int64  
 2   model         35622 non-null  object 
 3   cylinders     35622 non-null  int64  
 4   fuel          35622 non-null  object 
 5   odometer      35622 non-null  float64
 6   transmission  35622 non-null  object 
 7   type          35622 non-null  object 
 8   paint_color   35622 non-null  object 
 9   is_4wd        35622 non-null  int64  
 10  date_posted   35622 non-null  object 
 11  days_listed   35622 non-null  int64  
 12  condition     35622 non-null  object 
dtypes: float64(1), int64(5), object(7)
memory usage: 3.5+ MB


### cylinders

In [10]:
print(df_cd['cylinders'].describe())

print(f'\nvalores nulos = {df_cd["cylinders"].isna().sum()}')

df_cd = df_cd.dropna(subset=['cylinders'])

print(f'\nvalores nulos despues de limpiarlos = {df_cd["cylinders"].isna().sum()}')

df_cd['cylinders'] = df_cd['cylinders'].astype(int)
print(f'\nTipo de dato se cambia de float a int = {df_cd["cylinders"].dtype}')

count    35622.000000
mean         6.103840
std          1.636012
min          3.000000
25%          4.000000
50%          6.000000
75%          8.000000
max         12.000000
Name: cylinders, dtype: float64

valores nulos = 0

valores nulos despues de limpiarlos = 0

Tipo de dato se cambia de float a int = int64


### fuel

In [11]:
print(df_cd['fuel'].describe())

print(f'\nvalores nulos = {df_cd["fuel"].isna().sum()}')

valores_ordenados_f = sorted(pd.unique(df_cd['fuel']))
print(f'\n{valores_ordenados_f}')

conteo_other = (df_cd['fuel'] == 'other').sum()
print(f'\nEn la columna de fuel se repite "other" {conteo_other} veces')

df_cd = df_cd[df_cd['fuel'] != 'other']

valores_ordenados_f = sorted(pd.unique(df_cd['fuel']))
print(f'\n{valores_ordenados_f}')

conteo_other = (df_cd['fuel'] == 'other').sum()
print(f'\nEn la columna de fuel se repite "other" {conteo_other} veces')

print(f'\nEl tipo de columna es correcto = {df_cd['fuel'].dtype}')

count     35622
unique        4
top         gas
freq      32772
Name: fuel, dtype: object

valores nulos = 0

['diesel', 'electric', 'gas', 'hybrid']

En la columna de fuel se repite "other" 0 veces

['diesel', 'electric', 'gas', 'hybrid']

En la columna de fuel se repite "other" 0 veces

El tipo de columna es correcto = object


Se quitaron los valores de 'other' ya que representa poco y no da informacion

### odometer

In [12]:
print(df_cd['odometer'].describe())

print(f'\nvalores nulos = {df_cd["odometer"].isna().sum()}')

df_cd = df_cd.dropna(subset='odometer')

print(f'\nvalores nulos despues de limpiarlos = {df_cd["cylinders"].isna().sum()}')

print(f'\nEl tipo de columna es correcto = {df_cd['cylinders'].dtype}')

odometer_cero = df_cd[df_cd['odometer'] == 0].value_counts()

# Imprimir el DataFrame filtrado
print(odometer_cero)

print(f"\nCantidad de vehículos con odómetro en 0: {len(odometer_cero)}")

print(f'\nEl tipo de columna es correcto = {df_cd['odometer'].dtype}')

count     35622.000000
mean     117108.697097
std       64411.726935
min           1.000000
25%       72705.500000
50%      114473.500000
75%      156000.000000
max      990000.000000
Name: odometer, dtype: float64

valores nulos = 0

valores nulos despues de limpiarlos = 0

El tipo de columna es correcto = int64
Series([], Name: count, dtype: int64)

Cantidad de vehículos con odómetro en 0: 0

El tipo de columna es correcto = float64


fue aqui donde se detecto el error entre 'odometer' con valores altos y 'condicion' 'new' por lo que se quitaron los valores nulos desde el tratamiento de la columna 'condition' 

### transmission

In [13]:
print(df_cd['transmission'].describe())

print(f'\nvalores nulos = {df_cd["transmission"].isna().sum()}')

print('\n')
valores_ordenados_t = sorted(pd.unique(df_cd['transmission']))
print(valores_ordenados_t)

conteo_transmission = df_cd['transmission'].value_counts()
print(f'\nEl tipo de "transmission" "other" se repite {conteo_transmission.get('other', 0)} veces')

print(f'\nEl tipo de columna es correcto = {df_cd['transmission'].dtype}')

count         35622
unique            3
top       automatic
freq          32601
Name: transmission, dtype: object

valores nulos = 0


['automatic', 'manual', 'other']

El tipo de "transmission" "other" se repite 1049 veces

El tipo de columna es correcto = object


Tome la decision de dejar el valor 'other' en esta columna ya que si hay mas tipos de transmiciones.

### type

In [14]:
print(df_cd['type'].describe())

print(f'\nvalores nulos = {df_cd["type"].isna().sum()}')

valores_ordenados_ty = sorted(pd.unique(df_cd['type']))

print(f'\nEl tipo de columna es correcto = {df_cd['type'].dtype}')

count     35622
unique       13
top         SUV
freq       8567
Name: type, dtype: object

valores nulos = 0

El tipo de columna es correcto = object


se reviso la columna y todo parece en orden

### paint_color

In [15]:
print(df_cd['paint_color'].describe())

print(f'\nvalores nulos = {df_cd["paint_color"].isna().sum()}')

df_cd['paint_color'] = df_cd['paint_color'].fillna('Unknown')

print(f'\nvalores nulos despues de la limpieza = {df_cd["paint_color"].isna().sum()}')

valores_ordenados_p = sorted(pd.unique(df_cd['paint_color']))

print(f'\nEl tipo de columna es correcto = {df_cd['paint_color'].dtype}')

count     35622
unique       13
top       white
freq       6987
Name: paint_color, dtype: object

valores nulos = 0

valores nulos despues de la limpieza = 0

El tipo de columna es correcto = object


### is_4wd

In [16]:
print(df_cd['is_4wd'].describe())

print(f'\nvalores nulos = {df_cd["is_4wd"].isna().sum()}')

df_cd['is_4wd'] = df_cd['is_4wd'].fillna(0)

print(f'\nvalores nulos despues de la limpieza = {df_cd["paint_color"].isna().sum()}')

df_cd['is_4wd'] = df_cd['is_4wd'].astype(int)

print(f'\nEl tipo de columna se cambio de float a int = {df_cd['is_4wd'].dtype}')

count    35622.000000
mean         0.492617
std          0.499953
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          1.000000
Name: is_4wd, dtype: float64

valores nulos = 0

valores nulos despues de la limpieza = 0

El tipo de columna se cambio de float a int = int64




valores nulos = 18074

valores nulos despues de la limpieza = 0

El tipo de columna se cambio de float a int = int64


la columna tenia 18074 valores nulos pero al parecer es error de captura por lo que se cambio a 0 debido a que el valor capturado eraa de 1 que corresponde a que el vehículo tiene tracción en las cuatro ruedas (4WD).

### date_posted

In [17]:
print(df_cd['date_posted'].describe())

print(f'\nvalores nulos = {df_cd["date_posted"].isna().sum()}')

df_cd['date_posted'] = pd.to_datetime(df_cd['date_posted'], format='%Y-%m-%d')

print(f'\nEl tipo de columna se caambio de object a date correcto = {df_cd['date_posted'].dtype}')

count          35622
unique           354
top       2019-03-17
freq             128
Name: date_posted, dtype: object

valores nulos = 0

El tipo de columna se caambio de object a date correcto = datetime64[ns]


### days_listed

In [18]:
print(df_cd['days_listed'].describe())

print(f'\nvalores nulos = {df_cd["days_listed"].isna().sum()}')

valores_ordenados_d = sorted(pd.unique(df_cd['days_listed']))

dias_cero = df_cd[df_cd['days_listed'] == 0]

print(len(dias_cero))

dias_cero.sample(10)

count    35622.000000
mean        39.686289
std         28.132268
min          0.000000
25%         19.000000
50%         33.000000
75%         54.000000
max        271.000000
Name: days_listed, dtype: float64

valores nulos = 0
37


price  model_year                     model  cylinders    fuel  \
18483   9995        2006                  ram 1500          8     gas   
19196   2900        2009               honda civic          4     gas   
28971   1800        2002               honda civic          4     gas   
18693  10995        2010  chevrolet silverado 1500          8     gas   
1425   14000        1999                 ford f250          8  diesel   
11053   8200        2014              honda accord          4     gas   
20281   9450        2014               kia sorento          4     gas   
4380     500        2003              dodge dakota          6     gas   
13137  29999        2011            ford econoline         10     gas   
17419   9950        2008             toyota tundra          6     gas   

       odometer transmission    type paint_color  is_4wd date_posted  \
18483  101750.0    automatic   truck       black       1  2018-08-11   
19196  190222.0    automatic   sedan      silver       0  2018-12-17   
28971  238155.0    automatic   sedan      silver       0  2018-09-28   
18693  165178.0    automatic   truck        grey       1  2019-03-18   
1425   137500.0    automatic   truck         red       1  2018-09-27   
11053  121000.0    automatic   sedan      silver       0  2018-05-31   
20281   91776.0    automatic     SUV      silver       1  2018-08-27   
4380   215457.0    automatic  pickup     Unknown       0  2018-10-31   
13137  117000.0    automatic     van       white       0  2018-08-07   
17419  229000.0    automatic   truck         red       0  2019-04-16   

       days_listed condition  
18483            0      fair  
19196            0   salvage  
28971            0   salvage  
18693            0      fair  
1425             0      fair  
11053            0      fair  
20281            0      good  
4380             0   salvage  
13137            0      fair  
17419            0   salvage

In [19]:
df_cd

price  model_year           model  cylinders fuel  odometer  \
0       9400        2011          bmw x5          6  gas  145000.0   
1       5500        2013  hyundai sonata          4  gas  110000.0   
2      14900        2017    chrysler 200          4  gas   80903.0   
3      14990        2014    chrysler 300          6  gas   57954.0   
4      12990        2015    toyota camry          4  gas   79212.0   
...      ...         ...             ...        ...  ...       ...   
35617   3750        2005     ford taurus          6  gas  110200.0   
35618   9249        2013   nissan maxima          6  gas   88136.0   
35619   2700        2002     honda civic          4  gas  181500.0   
35620   3950        2009  hyundai sonata          4  gas  128000.0   
35621   7455        2013  toyota corolla          4  gas  139573.0   

      transmission   type paint_color  is_4wd date_posted  days_listed  \
0        automatic    SUV     Unknown       1  2018-06-23           19   
1        automatic  sedan         red       0  2019-02-07           79   
2        automatic  sedan       black       0  2019-04-02           28   
3        automatic  sedan       black       1  2018-06-20           15   
4        automatic  sedan       white       0  2018-12-27           73   
...            ...    ...         ...     ...         ...          ...   
35617    automatic  sedan      silver       0  2018-08-10           63   
35618    automatic  sedan       black       0  2018-10-03           37   
35619    automatic  sedan       white       0  2018-11-14           22   
35620    automatic  sedan        blue       0  2018-11-15           32   
35621    automatic  sedan       black       0  2018-07-02           71   

      condition  
0          fair  
1          fair  
2          good  
3          good  
4          good  
...         ...  
35617      fair  
35618      good  
35619   salvage  
35620      fair  
35621      fair  

[35622 rows x 13 columns]

In [22]:
df_cd.head()

price  model_year           model  cylinders fuel  odometer transmission  \
0   9400        2011          bmw x5          6  gas  145000.0    automatic   
1   5500        2013  hyundai sonata          4  gas  110000.0    automatic   
2  14900        2017    chrysler 200          4  gas   80903.0    automatic   
3  14990        2014    chrysler 300          6  gas   57954.0    automatic   
4  12990        2015    toyota camry          4  gas   79212.0    automatic   

    type paint_color  is_4wd date_posted  days_listed condition  
0    SUV     Unknown       1  2018-06-23           19      fair  
1  sedan         red       0  2019-02-07           79      fair  
2  sedan       black       0  2019-04-02           28      good  
3  sedan       black       1  2018-06-20           15      good  
4  sedan       white       0  2018-12-27           73      good